### PR TITLE
Linstor: add HA support and small cleanups

### PR DESCRIPTION
**Pull request by rp-**
### Description

This PR add HA support for Linstor primary storage.
It checks output of drbd event and drbd status output, to check if it has quorum or connections to other hosts,
which in turn reveals if this node is isolated or not.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [X] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Nested HA Linstor cluster with virtualBMC IPMI:
* Simulated node crash (kernel panic)
* Blocking of drbd incoming and outgoing connections

### CODE REVIEW

File: plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/util/LinstorUtil.java [hunk 1]
```diff
@@ -136,28 +136,33 @@ public static String getSnapshotPath(com.linbit.linstor.api.model.StoragePool sp
...
                 null,
                 null
-            );
+        );
+    }
+
+    public static long getCapacityBytes(String linstorUrl, String rscGroupName) {
+        DevelopersApi linstorApi = getLinstorAPI(linstorUrl);
+        try {
+            List<StoragePool> storagePools = getRscGroupStoragePools(linstorApi, rscGroupName);
```
Comment (sureshanaparti) [inline plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/util/LinstorUtil.java:165]: Hi @rp- _getRscGroupStoragePools()_ / other api method calls can return null? if so, in any case, pls handle them wherever applicable.

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/util/LinstorUtil.java (hunk 1 above)
Comment (rp-) [reply plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/util/LinstorUtil.java:165]: No `getRscGroupStoragePools` would rather throw an exception than returning `null`. I guess the worst that could come out of this is an empty List.

**Comment by DaanHoogland** [comment]
clgtm and i agree with [rohit](https://github.com/apache/cloudstack/pull/8407#pullrequestreview-1862186963), but can you expand a bit on the testing @rp-? (as there is 0% coverage in this PR)

**Comment by rp-** [comment]
> clgtm and i agree with [rohit](https://github.com/apache/cloudstack/pull/8407#pullrequestreview-1862186963), but can you expand a bit on the testing @rp-? (as there is 0% coverage in this PR)

Well as I said, I setup a nested cluster with a VBMC IPMI and configured everything in CloudStack and
tested a few failover scenarios. Like trigger a kernel panic on a node with a running VM (HA enabled) and let it fail over to another node and also see if it got fenced correctly.
The same with the node still seen by CloudStack, but blocking any DRBD traffic and having it look like the node can't access its storage anymore.
This will also trigger the fence action.
Other actions like VM starting/stoping/deleting worked as before.

Do you want me to test any specific scenarios?

**Review by DaanHoogland — APPROVED** [review]
clgtm, trouwens

**Comment by rp-** [comment]
I made a small change, to don't specify the full path to `hostname` to remedy concerns from this issue: https://github.com/apache/cloudstack/issues/8310

**Comment by rohityadavcloud** [comment]
The `hostname` issue is probably a critical/blocker for some users, @rp- can you consider a smaller fix for 4.18 branch? Thanks.

**Comment by rohityadavcloud** [comment]
@rp- could you help review/fix conflicts and change base branch if it's applicable for 4.18 or 4.19

**Comment by rp-** [comment]
> @rp- could you help review/fix conflicts and change base branch if it's applicable for 4.18 or 4.19

rebased to 4.19

**Comment by rp-** [comment]
@rohityadavcloud here https://github.com/apache/cloudstack/pull/8633 is a PR for 4.18

**Review by rohityadavcloud — APPROVED** [review]
LGTM - changes are strictly in the linstor plugin that our smoketests can't test. Github actions confirm other tests are passing with simulator.

**Comment by rohityadavcloud** [comment]
Hi @rp- @Philipp-Reisner quick question - does Linstor currently support VM HA with ACS 4.18/4.19? cc @harikrishna-patnala @borisstoyanov @NuxRo

**Comment by rp-** [comment]
@rohityadavcloud
Not in any released version, this patchset here is needed and also this fix here to get it working with larger setups:
https://github.com/apache/cloudstack/pull/8670

**Pull request merged**